pageserver/src/tenant/timeline.rs

```diff
@@ -2680,52 +2680,56 @@ impl Timeline {
-
-        // Calculate pitr cutoff point.
-        // If we cannot determine a cutoff LSN, be conservative and don't GC anything.
-        let mut pitr_cutoff_lsn: Lsn;
-
-        if pitr != Duration::ZERO {
-            // conservative, safe default is to remove nothing, when we have no
-            // commit timestamp data available
-            pitr_cutoff_lsn = *self.get_latest_gc_cutoff_lsn();
-
-            // First, calculate pitr_cutoff_timestamp and then convert it to LSN.
-            // If we don't have enough data to convert to LSN,
-            // play safe and don't remove any layers.
+        // First, calculate pitr_cutoff_timestamp and then convert it to LSN.
+        //
+        // Some unit tests depend on garbage-collection working even when
+        // CLOG data is missing, so that find_lsn_for_timestamp() doesn't
+        // work, so avoid calling it altogether if time-based retention is not
+        // configured. It would be pointless anyway.
+        let ts_cutoff: Option<LsnForTimestamp> = if pitr != Duration::ZERO {
             let now = SystemTime::now();
             if let Some(pitr_cutoff_timestamp) = now.checked_sub(pitr) {
                 let pitr_timestamp = to_pg_timestamp(pitr_cutoff_timestamp);
 
-                match self
-                    .find_lsn_for_timestamp(pitr_timestamp)
-                    .no_ondemand_download()?
-                {
-                    LsnForTimestamp::Present(lsn) => pitr_cutoff_lsn = lsn,
-                    LsnForTimestamp::Future(lsn) => {
-                        debug!("future({})", lsn);
-                        pitr_cutoff_lsn = gc_info.horizon_cutoff;
-                    }
-                    LsnForTimestamp::Past(lsn) => {
-                        debug!("past({})", lsn);
-                    }
-                    LsnForTimestamp::NoData(lsn) => {
-                        debug!("nodata({})", lsn);
-                    }
-                }
-                debug!("pitr_cutoff_lsn = {:?}", pitr_cutoff_lsn)
+                Some(
+                    self.find_lsn_for_timestamp(pitr_timestamp)
+                        .no_ondemand_download()?,
+                )
+            } else {
+                // If we don't have enough data to convert to LSN,
+                // play safe and don't remove any layers.
+                Some(LsnForTimestamp::NoData(Lsn(0)))
             }
         } else {
-            // No time-based retention. (Some unit tests depend on garbage-collection
-            // working even when CLOG data is missing, so that find_lsn_for_timestamp()
-            // above doesn't work.)
-            pitr_cutoff_lsn = gc_info.horizon_cutoff;
-        }
-        gc_info.pitr_cutoff = pitr_cutoff_lsn;
+            None
+        };
+
+        let mut gc_info = self.gc_info.write().unwrap();
+        gc_info.horizon_cutoff = cutoff_horizon;
+        gc_info.retain_lsns = retain_lsns;
+        gc_info.pitr_cutoff = match ts_cutoff {
```

Comment:
NIT: we could be more explicit (and less nested) for cases when no changes are needed

```
match ts_cutoff {
    None => {}, // No time-based retention was configured
    Some(LsnForTimestamp::Future(lsn)) => debug!("future({lsn})"),

    Some(LsnForTimestamp::Present(lsn)) => gc_info.horizon_cutoff = lsn,
    Some(LsnForTimestamp::Past(lsn)) => {
        debug!("past({lsn})");
        // conservative, safe default is to remove nothing, when we
        // have no commit timestamp data available
        gc_info.horizon_cutoff = *self.get_latest_gc_cutoff_lsn()
    }
    Some(LsnForTimestamp::NoData(lsn)) => {
        debug!("nodata({lsn})");
        // conservative, safe default is to remove nothing, when we
        // have no commit timestamp data available
        gc_info.horizon_cutoff = *self.get_latest_gc_cutoff_lsn()
    }
}
```

Reply:
Alternatively, can unpack this above, and return `Option<Lsn>` with the value to update the `gc_info.pitr_cutoff` with.

Reply:
> NIT: we could be more explicit (and less nested) for cases when no changes are needed
> 
> ```
> match ts_cutoff {
>     None => {}, // No time-based retention was configured
>     Some(LsnForTimestamp::Future(lsn)) => debug!("future({lsn})"),
> 
>     Some(LsnForTimestamp::Present(lsn)) => gc_info.horizon_cutoff = lsn,
>     Some(LsnForTimestamp::Past(lsn)) => {
>         debug!("past({lsn})");
>         // conservative, safe default is to remove nothing, when we
>         // have no commit timestamp data available
>         gc_info.horizon_cutoff = *self.get_latest_gc_cutoff_lsn()
>     }
>     Some(LsnForTimestamp::NoData(lsn)) => {
>         debug!("nodata({lsn})");
>         // conservative, safe default is to remove nothing, when we
>         // have no commit timestamp data available
>         gc_info.horizon_cutoff = *self.get_latest_gc_cutoff_lsn()
>     }
> }
> ```

Hmm, that doesn't do the same thing. With that, `pitr_cutoff` isn't updated at all in the None and Future cases. Currently, `pitr_cutoff` is updated to be equal to `horizon_cutoff`.

Reply:
>Alternatively, can unpack this above, and return Option<Lsn> with the value to update the gc_info.pitr_cutoff with.

Hmm, that might be more clear, agreed. I'll try that

Reply:
Refactored this one more time. Please take another look!